File: macos/Sources/ClaudeBrain/BrainInspector.swift
import SwiftUI

/// Estado de instalación de una pieza del cerebro, leído de la realidad (`~/.claude`).
enum BrainStatus {
    case installed        // global: script presente + cableado, o norma/skill presente
    case presentNotWired  // global: el script existe pero NO está cableado en settings.json
    case absent           // global: se esperaba y no está
    case repoScoped       // viaja por repo (no verificable desde el ~/.claude global)

    // NOTA: mantenemos los 4 casos para la PRECISIÓN interna (el detalle al tocar los distingue por
    // `label`), pero de cara al usuario el color/símbolo COLAPSAN a binario: verde=bien, rojo=faltante
    // (sin cablear y ausente se ven igual → "algo falta, cúralo"), azul discreto=por-repo (no cuenta).
    var symbol: String {
        switch self {
        case .installed:       return "checkmark.circle.fill"
        case .presentNotWired: return "exclamationmark.circle.fill"
        case .absent:          return "exclamationmark.circle.fill"
        case .repoScoped:      return "circle.dashed"
        }
    }
    var color: Color {
        switch self {
        case .installed:       return Color(hex: "#3aa76d")
        case .presentNotWired: return Color(hex: "#dc3545")
        case .absent:          return Color(hex: "#dc3545")
        case .repoScoped:      return Color(hex: "#4a90d9").opacity(0.6)
        }
    }
    /// Texto de ayuda al expandir la pieza.
    var label: String {
        switch self {
        case .installed:       return "instalado + cableado en tu ~/.claude"
        case .presentNotWired: return "el script existe pero NO está cableado en settings.json"
        case .absent:          return "no instalado en tu ~/.claude"
        case .repoScoped:      return "viaja por repo: se copia al .claude/ de cada proyecto"
        }
    }
}

/// Evidencia real leída de `~/.claude` en un instante dado. Puro I/O de lectura, tolerante a fallos.
struct BrainState {
    var presentHooks: Set<String> = []   // basenames sin .sh en ~/.claude/hooks
    var wiredHooks: Set<String> = []      // basenames referenciados en settings.json
    var hasNorms: Bool = false            // ~/.claude/CLAUDE.md trae el marcador BEGIN claude-brain
    var skills: Set<String> = []          // subcarpetas de ~/.claude/skills con SKILL.md
    var extras: [String] = []             // hooks cableados que no están en el catálogo conocido
    var version: String? = nil            // sello de ~/.claude/.brain-version línea 1 (install-brain.sh); nil si no está
    var versionDate: String? = nil        // fecha de instalación (línea 2 del sello); nil en sellos viejos de 1 línea
    var scannedAt: Date = Date()

    /// Hooks de tier {global, both} (kind=hook) que install-brain.sh cablea en el ~/.claude global.
    /// DEBE coincidir con brain/hooks/MANIFEST {global,both}; lo verifica el drift-check del widget
    /// (test-brain.sh, chequeo e-widget). Al agregar/mover un hook en el MANIFEST, refleja aquí.
    static let knownGlobalHooks: Set<String> = [
        "git-branch-guard", "merge-squash-guard", "confirmar-merge-develop",
        "recordar-dashboard", "secret-scan", "rama-vieja", "proteger-arbol",
        "limite-gasto", "delegacion-gate", "delegacion-registrar", "delegacion-reporte",
        "rehidratar-hilo", "aviso-contexto", "aviso-drift-cerebro", "barrer-ramas",
        "entorno-maquina-guard",
    ]
    /// Hooks de tier repo (kind=hook): viajan por-repo, se cargan si la sesión INICIA en el repo.
    /// DEBE coincidir con brain/hooks/MANIFEST {repo} (lo verifica test-brain.sh).
    static let knownRepoHooks: Set<String> = [
        "sesion-inicio", "dod-verificar", "recordar-cosechar", "recordar-unificar-cerebro",
    ]

    /// # de piezas GLOBALES esperadas que FALTAN: hooks no (presentes+cableados) + normas + la skill.
    /// Espeja el criterio del recuadro de salud; alimenta el 🩹 del riel y de la píldora de la barra.
    var globalMissing: Int {
        var n = 0
        for h in Self.knownGlobalHooks where !(presentHooks.contains(h) && wiredHooks.contains(h)) { n += 1 }
        if !hasNorms { n += 1 }
        if !skills.contains("cerrar-slice") { n += 1 }
        return n
    }
    var isComplete: Bool { globalMissing == 0 }
}

/// Lee `~/.claude` y arma un `BrainState`. Todo el I/O es de LECTURA y fail-safe (si algo falta,
/// esa pieza queda como ausente en vez de romper). Así la pestaña Cerebro refleja la realidad.
enum BrainInspector {
    static func inspect() -> BrainState {
        var st = BrainState()
        let fm = FileManager.default
        let home = fm.homeDirectoryForCurrentUser
        let claude = home.appendingPathComponent(".claude")

        // (1) Hooks presentes: *.sh en ~/.claude/hooks
        let hooksDir = claude.appendingPathComponent("hooks")
        if let items = try? fm.contentsOfDirectory(atPath: hooksDir.path) {
            for f in items where f.hasSuffix(".sh") {
                st.presentHooks.insert(String(f.dropLast(3)))
            }
        }

        // (2) Hooks cableados: basenames referenciados en los comandos de settings.json
        let settings = claude.appendingPathComponent("settings.json")
        if let data = try? Data(contentsOf: settings),
           let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let hooks = root["hooks"] as? [String: Any] {
            var commands: [String] = []
            for (_, entriesAny) in hooks {
                guard let entries = entriesAny as? [[String: Any]] else { continue }
                for entry in entries {
                    guard let hs = entry["hooks"] as? [[String: Any]] else { continue }
                    for h in hs { if let c = h["command"] as? String { commands.append(c) } }
                }
            }
            let rx = try? NSRegularExpression(pattern: #"/hooks/([A-Za-z0-9._-]+)\.sh"#)
            for c in commands {
                let range = NSRange(c.startIndex..., in: c)
                rx?.enumerateMatches(in: c, range: range) { m, _, _ in
                    if let m, let r = Range(m.range(at: 1), in: c) {
                        st.wiredHooks.insert(String(c[r]))
                    }
                }
            }
        }

        // (3) Normas: instaladas por el marcador de inyección, O presentes por contenido (escritas a
        //     mano). Ambas gobiernan de verdad → cuentan como activas (doc=realidad: reflejamos el
        //     efecto real, no solo si pasaron por install-brain.sh).
        let claudeMd = claude.appendingPathComponent("CLAUDE.md")
        if let txt = try? String(contentsOf: claudeMd, encoding: .utf8) {
            st.hasNorms = txt.contains("BEGIN claude-brain")
                || txt.contains("Definición de \"LISTO\"")
                || txt.contains("reflejo de la realidad")
        }

        // (4) Skills: subcarpetas de ~/.claude/skills con un SKILL.md
        let skillsDir = claude.appendingPathComponent("skills")
        if let dirs = try? fm.contentsOfDirectory(atPath: skillsDir.path) {
            for d in dirs {
                let skillFile = skillsDir.appendingPathComponent(d).appendingPathComponent("SKILL.md")
                if fm.fileExists(atPath: skillFile.path) { st.skills.insert(d) }
            }
        }

        // (5) Versión instalada: sello ~/.claude/.brain-version (lo estampa install-brain.sh). Contrato
        //     de 2 líneas: L1 = versión "<PREFIJO>.<count>", L2 = fecha "YYYY-MM-DD". Back-compat de
        //     LECTURA: si el sello trae 1 sola línea (formato viejo) → versión = L1, fecha = nil.
        //     Fail-safe: ausente/vacío → nil y la UI simplemente no muestra versión.
        let versionFile = claude.appendingPathComponent(".brain-version")
        if let raw = try? String(contentsOf: versionFile, encoding: .utf8) {
            let lines = raw.split(separator: "\n", omittingEmptySubsequences: false)
                .map { $0.trimmingCharacters(in: .whitespaces) }
            if let v = lines.first, !v.isEmpty { st.version = v }
            if lines.count >= 2, !lines[1].isEmpty { st.versionDate = lines[1] }
        }

        // (6) Extras: hooks cableados que no reconocemos (ni global ni repo-scoped del catálogo)
        let known = BrainState.knownGlobalHooks.union(BrainState.knownRepoHooks)
        st.extras = st.wiredHooks.subtracting(known).sorted()

        st.scannedAt = Date()
        return st
    }
}

File: windows/src/Cortex/BrainInspector.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Cortex;

/// <summary>
/// Estado de instalación de una pieza del cerebro, leído de la realidad (`~/.claude`).
/// Espejo 1:1 del enum `BrainStatus` de la GUI macOS (BrainInspector.swift).
/// </summary>
public enum BrainStatus
{
    Installed,        // global: script presente + cableado, o norma/skill presente
    PresentNotWired,  // global: el script existe pero NO está cableado en settings.json
    Absent,           // global: se esperaba y no está
    RepoScoped,       // viaja por repo (no verificable desde el ~/.claude global)
}

/// <summary>
/// Evidencia real leída de `~/.claude` en un instante dado. Puro I/O de lectura, tolerante a
/// fallos. Espejo del struct `BrainState` de Swift.
/// </summary>
public sealed class BrainState
{
    public HashSet<string> PresentHooks { get; } = new();   // basenames sin .sh en ~/.claude/hooks
    public HashSet<string> WiredHooks { get; } = new();     // basenames referenciados en settings.json
    public bool HasNorms { get; set; }                      // ~/.claude/CLAUDE.md trae marcador/normas
    public HashSet<string> Skills { get; } = new();         // subcarpetas de ~/.claude/skills con SKILL.md
    public List<string> Extras { get; set; } = new();       // hooks cableados fuera del catálogo
    public string? Version { get; set; }                    // sello ~/.claude/.brain-version (línea 1); null si no está
    public string? VersionDate { get; set; }                // fecha del sello (línea 2, YYYY-MM-DD); null en formato viejo de 1 línea
    public DateTime ScannedAt { get; set; } = DateTime.Now;

    /// Hooks de tier {global, both} (kind=hook) que install-brain.sh cablea en el ~/.claude global.
    /// DEBE coincidir con brain/hooks/MANIFEST {global,both}; lo verifica el drift-check del widget
    /// (test-brain.sh, chequeo e-widget). Al agregar/mover un hook en el MANIFEST, refleja aquí.
    public static readonly HashSet<string> KnownGlobalHooks = new()
    {
        "git-branch-guard", "merge-squash-guard", "confirmar-merge-develop",
        "recordar-dashboard", "secret-scan", "rama-vieja", "proteger-arbol",
        "proteger-fuente-cerebro",
        "limite-gasto", "delegacion-gate", "delegacion-registrar", "delegacion-reporte", "recordar-orquestar",
        "rehidratar-hilo", "aviso-contexto", "aviso-drift-cerebro", "hud-stale", "exportar-sesion-master", "barrer-ramas",
        "entorno-maquina-guard", "no-bypass-deploy",
    };

    /// Hooks de tier repo (kind=hook): viajan por-repo, se cargan si la sesión INICIA en el repo.
    /// DEBE coincidir con brain/hooks/MANIFEST {repo} (lo verifica test-brain.sh).
    public static readonly HashSet<string> KnownRepoHooks = new()
    {
        "sesion-inicio", "dod-verificar", "recordar-cosechar", "recordar-unificar-cerebro",
    };

    /// Estado real de una pieza (por nombre) contra la evidencia leída. Espejo de `status(_:_:)` de Swift.
    public BrainStatus StatusOf(string name)
    {
        if (KnownGlobalHooks.Contains(name))
        {
            bool p = PresentHooks.Contains(name), w = WiredHooks.Contains(name);
            return p && w ? BrainStatus.Installed : (p ? BrainStatus.PresentNotWired : BrainStatus.Absent);
        }
        if (KnownRepoHooks.Contains(name)) return BrainStatus.RepoScoped;
        return name switch
        {
            "cerrar-slice" or "checkpoint" or "to-do" or "diagramar" or "auditar-proceso-algoritmo"
                or "auditar-coherencia-cerebro" or "auditar-suficiencia-operativa" or "auditor-semantico" or "consolidar-cerebro" or "canonizar-cerebro" or "desinflar-memorias"
                or "orquestar-fanout" or "turno-nocturno" or "cosechar-sesion" or "unificar-cerebro"
                or "investigar-dominio" or "construir-missing-manual" or "positivar-doc" or "revisar-entregables-agentes"
                or "zoom-screenshot" or "claude-proyecto-autocontenido" or "reubicar-master"
                or "ingenieria-inversa-gui-db-navegador" or "markdown-a-pdf"
                => Skills.Contains(name) ? BrainStatus.Installed : BrainStatus.Absent,
            "Definition of Done" or "Doc <= realidad" or "Flujo de git" or "Costo de delegación"
                => HasNorms ? BrainStatus.Installed : BrainStatus.Absent,
            _ => BrainStatus.Absent,
        };
    }
}

/// <summary>
/// Lee `~/.claude` y arma un `BrainState`. Todo el I/O es de LECTURA y fail-safe (si algo falta,
/// esa pieza queda como ausente en vez de romper). Así la pestaña Cerebro refleja la realidad.
/// Réplica en .NET de `BrainInspector.inspect()` (Swift). En Windows el ~/.claude vive en el HOME
/// del usuario (UserProfile), así que usamos esa carpeta.
/// </summary>
public static class BrainInspector
{
    // Regex: basenames referenciados como /hooks/<nombre>.sh en los comandos de settings.json.
    private static readonly Regex HookRefRx =
        new(@"/hooks/([A-Za-z0-9._-]+)\.sh", RegexOptions.Compiled);

    public static BrainState Inspect()
    {
        var st = new BrainState();
        string home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        string claude = Path.Combine(home, ".claude");

        // (1) Hooks presentes: *.sh en ~/.claude/hooks
        try
        {
            string hooksDir = Path.Combine(claude, "hooks");
            if (Directory.Exists(hooksDir))
                foreach (var f in Directory.EnumerateFiles(hooksDir, "*.sh"))
                    st.PresentHooks.Add(Path.GetFileNameWithoutExtension(f));
        }
        catch { /* fail-safe: sin hooks presentes */ }

        // (2) Hooks cableados: basenames referenciados en los comandos de settings.json
        try
        {
            string settings = Path.Combine(claude, "settings.json");
            if (File.Exists(settings))
            {
                using var doc = JsonDocument.Parse(File.ReadAllText(settings, Encoding.UTF8));
                if (doc.RootElement.ValueKind == JsonValueKind.Object &&
                    doc.RootElement.TryGetProperty("hooks", out var hooks) &&
                    hooks.ValueKind == JsonValueKind.Object)
                {
                    foreach (var evt in hooks.EnumerateObject())
                    {
                        if (evt.Value.ValueKind != JsonValueKind.Array) continue;
                        foreach (var entry in evt.Value.EnumerateArray())
                        {
                            if (entry.ValueKind != JsonValueKind.Object ||
                                !entry.TryGetProperty("hooks", out var hs) ||
                                hs.ValueKind != JsonValueKind.Array) continue;
                            foreach (var h in hs.EnumerateArray())
                            {
                                if (h.ValueKind == JsonValueKind.Object &&
                                    h.TryGetProperty("command", out var cmd) &&
                                    cmd.ValueKind == JsonValueKind.String)
                                {
                                    foreach (Match m in HookRefRx.Matches(cmd.GetString() ?? ""))
                                        st.WiredHooks.Add(m.Groups[1].Value);
                                }
                            }
                        }
                    }
                }
            }
        }
        catch { /* fail-safe: sin cableado */ }

        // (3) Normas: instaladas por marcador de inyección, O presentes por contenido (escritas a
        //     mano). Ambas gobiernan de verdad → cuentan como activas (doc=realidad).
        try
        {
            string claudeMd = Path.Combine(claude, "CLAUDE.md");
            if (File.Exists(claudeMd))
            {
                string txt = File.ReadAllText(claudeMd, Encoding.UTF8);
                st.HasNorms = txt.Contains("BEGIN cortex")
                    || txt.Contains("Definición de \"LISTO\"")
                    || txt.Contains("reflejo de la realidad");
            }
        }
        catch { /* fail-safe: sin normas */ }

        // (4) Skills: subcarpetas de ~/.claude/skills con un SKILL.md
        try
        {
            string skillsDir = Path.Combine(claude, "skills");
            if (Directory.Exists(skillsDir))
                foreach (var d in Directory.EnumerateDirectories(skillsDir))
                    if (File.Exists(Path.Combine(d, "SKILL.md")))
                        st.Skills.Add(Path.GetFileName(d));
        }
        catch { /* fail-safe: sin skills */ }

        // (5) Versión instalada del brain: sello ~/.claude/.brain-version (lo estampa
        //     install-brain.sh copiando brain/VERSION). Formato NUEVO = 2 líneas: línea 1 = versión,
        //     línea 2 = fecha YYYY-MM-DD. Back-compat: si viene 1 sola línea (formato viejo) →
        //     versión = esa línea, fecha = null. Fail-safe: ausente/vacío → null (instalación previa
        //     al sello) y la UI simplemente no muestra versión.
        try
        {
            string verFile = Path.Combine(claude, ".brain-version");
            if (File.Exists(verFile))
            {
                var lines = File.ReadAllLines(verFile, Encoding.UTF8);
                if (lines.Length >= 1 && lines[0].Trim().Length > 0) st.Version = lines[0].Trim();
                if (lines.Length >= 2 && lines[1].Trim().Length > 0) st.VersionDate = lines[1].Trim();
            }
        }
        catch { /* fail-safe: sin versión */ }

        // (6) Extras: hooks cableados que no reconocemos (ni global ni repo-scoped del catálogo)
        var known = new HashSet<string>(BrainState.KnownGlobalHooks);
        known.UnionWith(BrainState.KnownRepoHooks);
        st.Extras = st.WiredHooks.Where(h => !known.Contains(h)).OrderBy(h => h).ToList();

        st.ScannedAt = DateTime.Now;
        return st;
    }

    // ── Presentación de cada estado (glifo + etiqueta). El color lo resuelve PopupForm. ──
    //
    // De cara al usuario el glifo COLAPSA a binario (espejo de `BrainStatus.symbol` en macOS):
    // ✓ verde = bien; ! rojo = falta algo (sin cablear y ausente se ven igual → "cúralo");
    // ○ azul discreto = por-repo (no cuenta). El matiz fino de los 4 estados sigue en `Label`.

    /// Glifo binario de punto de estado. ✓ instalado · ! atención (sin cablear/ausente) · ○ por-repo.
    public static string Glyph(BrainStatus s) => s switch
    {
        BrainStatus.Installed => "✓",
        BrainStatus.PresentNotWired => "!",
        BrainStatus.Absent => "!",
        BrainStatus.RepoScoped => "○",
        _ => "!",
    };

    /// Texto de ayuda al expandir la pieza (espejo de `label` en Swift).
    public static string Label(BrainStatus s) => s switch
    {
        BrainStatus.Installed => "instalado + cableado en tu ~/.claude",
        BrainStatus.PresentNotWired => "el script existe pero NO está cableado en settings.json",
        BrainStatus.Absent => "no instalado en tu ~/.claude",
        BrainStatus.RepoScoped => "viaja por repo: se copia al .claude/ de cada proyecto",
        _ => "",
    };
}
